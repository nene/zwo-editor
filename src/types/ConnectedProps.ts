type AnyFn = (...args: any) => any;

type ReturnVoid<F extends AnyFn> = (...args: Parameters<F>) => void;

type FunctionMap = { [key: string]: AnyFn };

type VoidFunctionMap<T extends FunctionMap> = {
  [P in keyof T]: ReturnVoid<T[P]>;
};

// Derives Props type from mapStateToProps and mapDispatchToProps
export type ConnectedProps<
  MapStateToProps extends AnyFn,
  MapDispatchToProps extends FunctionMap = {}
> = ReturnType<MapStateToProps> & VoidFunctionMap<MapDispatchToProps>;
