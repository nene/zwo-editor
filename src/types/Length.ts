export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;

// Alternative approach
//
// This way we could avoid storing class instances to Redux,
// which eliminates problems with serialization & deserialization.
/*
export type Distance = { meters: number };
export type Duration = { seconds: number };
export type Length = Distance | Duration;

export const Distance = (meters: number) => ({ meters });
export const Duration = (seconds: number) => ({ seconds });

export const isDistance = (obj: Length): obj is Distance => 'meters' in obj;
export const isDuration = (obj: Length): obj is Duration => 'seconds' in obj;
*/
