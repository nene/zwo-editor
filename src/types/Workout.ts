import { SportType } from "./SportType";
import { Interval } from "./Interval";
import { LengthType } from "./LengthType";

export interface Workout {
  author: string;
  name: string;
  description: string;
  sportType: SportType;
  lengthType: LengthType;
  tags: string[];
  intervals: Array<Interval>;
}

export function createEmptyWorkout(
  sportType: SportType,
  lengthType: LengthType
): Workout {
  return {
    author: "",
    name: "",
    description: "",
    sportType: sportType,
    lengthType: lengthType,
    tags: [],
    intervals: [],
  };
}
