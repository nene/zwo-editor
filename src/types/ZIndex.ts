export enum ZIndex {
  fader = 1,
  selected = 5,
  instruction = 11,
  tooltip = 100,
  popup = 1000,
}
