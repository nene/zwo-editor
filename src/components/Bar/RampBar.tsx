import React, { useState } from "react";
import styled from "styled-components";
import { zoneColor, ZoneColor, Zones, ZonesArray } from "../../types/Zones";
import { Resizable } from "re-resizable";
import Tooltip from "../Tooltip/Tooltip";
import { RampInterval } from "../../types/Interval";
import { intensityMultiplier } from "./multipliers";
import { WorkoutMode } from "../../modes/WorkoutMode";
import { BarIcons } from "./BarIcons";

interface IDictionary {
  [index: string]: number;
}

interface RampBarProps {
  interval: RampInterval;
  mode: WorkoutMode;
  onChange: (interval: RampInterval) => void;
  onClick: (id: string) => void;
  selected: boolean;
}

const RampBar = ({ interval, mode, ...props }: RampBarProps) => {
  const [showTooltip, setShowTooltip] = useState(false);

  const handleCadenceChange = (cadence: number | undefined) => {
    props.onChange({ ...interval, cadence: cadence });
  };

  const width = mode.lengthToWidth(interval.length);
  const [resizableWidth, setResizableWidth] = useState(width);

  const startHeight = mode.intensityToHeight(interval.startIntensity);
  const [resizableStartHeight, setResizableStartHeight] = useState(startHeight);

  const endHeight = mode.intensityToHeight(interval.endIntensity);
  const [resizableEndHeight, setResizableEndHeight] = useState(endHeight);

  const handleStartResizeStop = (dHeight: number) => {
    setResizableStartHeight(resizableStartHeight + dHeight);
  };
  const handleEndResizeStop = (dWidth: number, dHeight: number) => {
    setResizableWidth(resizableWidth + dWidth);
    setResizableEndHeight(resizableEndHeight + dHeight);
  };

  const handleStartResize = (dHeight: number) => {
    props.onChange({
      ...interval,
      startIntensity: mode.heightToIntensity(resizableStartHeight + dHeight),
    });
  };
  const handleEndResize = (dWidth: number, dHeight: number) => {
    props.onChange({
      ...interval,
      length: mode.widthToLength(resizableWidth + dWidth),
      startIntensity: mode.heightToIntensity(resizableStartHeight),
      endIntensity: mode.heightToIntensity(resizableEndHeight + dHeight),
    });
  };

  return (
    <Container
      onMouseEnter={() => setShowTooltip(true)}
      onMouseLeave={() => setShowTooltip(false)}
      style={props.selected ? { zIndex: 1 } : {}}
      onClick={() => props.onClick(interval.id)}
    >
      {(props.selected || showTooltip) && (
        <Tooltip
          interval={interval}
          mode={mode}
          onCadenceChange={(cadence) => handleCadenceChange(cadence)}
        />
      )}
      <Ramp onClick={() => props.onClick(interval.id)}>
        <ResizableRamp
          size={{
            width: resizableWidth / 2,
            height: resizableStartHeight,
          }}
          minWidth={3}
          minHeight={intensityMultiplier * Zones.Z1.min}
          maxHeight={intensityMultiplier * Zones.Z6.max}
          enable={{ top: true, right: true }}
          grid={[1, 1]}
          onResizeStop={(e, direction, ref, d) =>
            handleStartResizeStop(d.height)
          }
          onResize={(e, direction, ref, d) => handleStartResize(d.height)}
        >
          <BarIcons
            height={startHeight > endHeight ? endHeight : startHeight}
            cadence={interval.cadence}
            instructions={interval.instructions}
          />
        </ResizableRamp>
        <ResizableRamp
          size={{
            width: resizableWidth / 2,
            height: resizableEndHeight,
          }}
          minWidth={3}
          minHeight={intensityMultiplier * Zones.Z1.min}
          maxHeight={intensityMultiplier * Zones.Z6.max}
          enable={{ top: true, right: true }}
          grid={[1, 1]}
          onResizeStop={(e, direction, ref, d) =>
            handleEndResizeStop(d.width, d.height)
          }
          onResize={(e, direction, ref, d) =>
            handleEndResize(d.width, d.height)
          }
        ></ResizableRamp>
      </Ramp>
      <Rainbow
        interval={interval}
        startHeight={startHeight}
        endHeight={endHeight}
      />
      <SvgPolygons
        width={width}
        startHeight={startHeight}
        endHeight={endHeight}
      />
    </Container>
  );
};

const Container = styled.div`
  position: relative;
`;

const Ramp = styled.div`
  display: flex;
  align-items: flex-end;
`;

const ResizableRamp = styled(Resizable)`
  border-top: 1px dotted gray;
  z-index: 5;
`;

const Rainbow: React.FC<{
  interval: RampInterval;
  startHeight: number;
  endHeight: number;
}> = ({ interval, startHeight, endHeight }) => {
  const trapezeHeight = endHeight > startHeight ? endHeight : startHeight;
  const bars =
    endHeight > startHeight
      ? calculateColors(interval.startIntensity, interval.endIntensity)
      : calculateColors(interval.endIntensity, interval.startIntensity);
  const flexDirection = endHeight > startHeight ? "row" : "row-reverse";

  return (
    <RampColors
      style={{
        height: trapezeHeight,
        flexDirection: flexDirection,
        backgroundColor: zoneColor(interval.startIntensity),
      }}
    >
      <Color
        style={{
          backgroundColor: ZoneColor.GRAY,
          width: `${
            (bars["Z1"] * 100) /
            Math.abs(interval.endIntensity - interval.startIntensity)
          }%`,
        }}
      />
      <Color
        style={{
          backgroundColor: ZoneColor.BLUE,
          width: `${
            (bars["Z2"] * 100) /
            Math.abs(interval.endIntensity - interval.startIntensity)
          }%`,
        }}
      />
      <Color
        style={{
          backgroundColor: ZoneColor.GREEN,
          width: `${
            (bars["Z3"] * 100) /
            Math.abs(interval.endIntensity - interval.startIntensity)
          }%`,
        }}
      />
      <Color
        style={{
          backgroundColor: ZoneColor.YELLOW,
          width: `${
            (bars["Z4"] * 100) /
            Math.abs(interval.endIntensity - interval.startIntensity)
          }%`,
        }}
      />
      <Color
        style={{
          backgroundColor: ZoneColor.ORANGE,
          width: `${
            (bars["Z5"] * 100) /
            Math.abs(interval.endIntensity - interval.startIntensity)
          }%`,
        }}
      />
      <Color
        style={{
          backgroundColor: ZoneColor.RED,
          width: `${
            (bars["Z6"] * 100) /
            Math.abs(interval.endIntensity - interval.startIntensity)
          }%`,
        }}
      />
    </RampColors>
  );
};

const RampColors = styled.div`
  position: absolute;
  bottom: 0px;
  left: 0;
  width: 100%;
  display: flex;
`;

const Color = styled.div`
  &:first-child {
    border-bottom-left-radius: 5px;
  }
  &:last-child {
    border-bottom-right-radius: 5px;
  }
`;

function calculateColors(start: number, end: number): IDictionary {
  const bars = {} as IDictionary;

  ZonesArray.forEach((zone, index) => {
    if (start >= zone[0] && start < zone[1]) {
      bars["Z" + (index + 1)] = zone[1] - start;
    } else if (end >= zone[0] && end < zone[1]) {
      bars["Z" + (index + 1)] = end - zone[0];
    } else if (end >= zone[1] && start < zone[0]) {
      bars["Z" + (index + 1)] = zone[1] - zone[0];
    } else {
      bars["Z" + (index + 1)] = 0;
    }
  });
  return bars;
}

const SvgPolygons: React.FC<{
  width: number;
  startHeight: number;
  endHeight: number;
}> = ({ width, startHeight, endHeight }) => {
  const trapezeHeight = endHeight > startHeight ? endHeight : startHeight;
  const trapezeTop =
    endHeight > startHeight ? endHeight - startHeight : startHeight - endHeight;

  const vertexB = endHeight > startHeight ? 0 : width;
  const vertexA = endHeight > startHeight ? trapezeTop : 0;
  const vertexD = endHeight > startHeight ? 0 : trapezeTop;

  return (
    <RampSvg height={`${trapezeHeight}`} width={`${width}`}>
      <TransparentPolygon
        points={`0,${vertexA} 0,${trapezeHeight} ${width},${trapezeHeight} ${width},${vertexD}`}
      />
      <HiddenPolygon points={`0,0 ${vertexB},${trapezeTop} ${width},0`} />
    </RampSvg>
  );
};

const RampSvg = styled.svg`
  position: absolute;
  bottom: 0;
  left: 0;
`;

const TransparentPolygon = styled.polygon`
  fill: transparent;
`;

const HiddenPolygon = styled.polygon`
  fill: white;
  display: hidden;
`;

export default RampBar;
