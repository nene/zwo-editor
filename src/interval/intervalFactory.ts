import { v4 as uuidv4 } from "uuid";
import { PaceType } from "../types/PaceType";
import {
  FreeInterval,
  Interval,
  RampInterval,
  RepetitionInterval,
  SteadyInterval,
} from "../types/Interval";
import { Distance, Duration, Length } from "../types/Length";
import { WorkoutMode } from "../modes/WorkoutMode";

const defaultDuration = Duration(300);
const defaultDistance = Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: "steady",
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: "ramp",
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: "free",
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: undefined,
      instructions: [],
      ...interval,
    };
  },

  repetition(
    interval: Partial<RepetitionInterval>,
    mode: WorkoutMode
  ): RepetitionInterval {
    return {
      type: "repetition",
      id: uuidv4(),
      repeat: 3,
      onLength: mode.lengthType === "time" ? Duration(30) : Distance(200),
      offLength: mode.lengthType === "time" ? Duration(120) : Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      onCadence: undefined,
      offCadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
