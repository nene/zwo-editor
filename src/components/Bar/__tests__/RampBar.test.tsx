import React from "react";
import RampBar from "../RampBar";
import { Zones } from "../../../types/Zones";
import renderer from "react-test-renderer";
import intervalFactory from "../../../interval/intervalFactory";
import createMode from "../../../modes/createMode";
import { Duration } from "../../../types/Length";

const MockReact = React;

jest.mock("../../Tooltip/Tooltip", () => (props: any) =>
  MockReact.createElement("Tooltip", props)
);

jest.mock("re-resizable", () => ({
  Resizable: (props: any) => MockReact.createElement("Resizable", props),
}));

jest.mock("uuid", () => ({
  v4: () => `mock-id`,
}));

describe("<SteadyBar>", () => {
  it("renders", () => {
    const mode = createMode({
      sportType: "bike",
      ftp: 250,
      weight: 75,
      runningTimes: [],
      lengthType: "time",
    });
    const interval = intervalFactory.ramp(
      {
        length: new Duration(50),
        startIntensity: Zones.Z2.min,
        endIntensity: Zones.Z4.min,
      },
      mode
    );

    const component = renderer.create(
      <RampBar
        interval={interval}
        mode={mode}
        selected={false}
        onChange={() => {}}
        onClick={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("renders with Tooltip when selected", () => {
    const mode = createMode({
      sportType: "bike",
      ftp: 250,
      weight: 75,
      runningTimes: [],
      lengthType: "time",
    });
    const interval = intervalFactory.ramp(
      {
        length: new Duration(50),
        startIntensity: Zones.Z2.min,
        endIntensity: Zones.Z4.min,
      },
      mode
    );

    const component = renderer.create(
      <RampBar
        interval={interval}
        mode={mode}
        selected={true}
        onChange={() => {}}
        onClick={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
