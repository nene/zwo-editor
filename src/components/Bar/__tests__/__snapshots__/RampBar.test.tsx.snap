// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SteadyBar> renders 1`] = `
.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c2 {
  border-top: 1px dotted gray;
  z-index: 5;
}

.c3 {
  position: absolute;
  bottom: 0px;
  left: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4:first-child {
  border-bottom-left-radius: 5px;
}

.c4:last-child {
  border-bottom-right-radius: 5px;
}

.c5 {
  position: absolute;
  bottom: 0;
  left: 0;
}

.c6 {
  fill: transparent;
}

.c7 {
  fill: white;
  display: hidden;
}

<div
  className="c0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="c1"
    onClick={[Function]}
  >
    <Resizable
      className="c2"
      enable={
        Object {
          "right": true,
          "top": true,
        }
      }
      grid={
        Array [
          1,
          1,
        ]
      }
      maxHeight={500}
      minHeight={25}
      minWidth={3}
      onResize={[Function]}
      onResizeStop={[Function]}
      size={
        Object {
          "height": 151.25,
          "width": 8.333333333333334,
        }
      }
    />
    <Resizable
      className="c2"
      enable={
        Object {
          "right": true,
          "top": true,
        }
      }
      grid={
        Array [
          1,
          1,
        ]
      }
      maxHeight={500}
      minHeight={25}
      minWidth={3}
      onResize={[Function]}
      onResizeStop={[Function]}
      size={
        Object {
          "height": 225.25,
          "width": 8.333333333333334,
        }
      }
    />
  </div>
  <div
    className="c3"
    style={
      Object {
        "backgroundColor": "#0E90D4",
        "flexDirection": "row",
        "height": 225.25,
      }
    }
  >
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#807F80",
          "width": "0%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#0E90D4",
          "width": "53.37837837837838%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#00C46A",
          "width": "46.62162162162162%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#FFCB00",
          "width": "0%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#FF6430",
          "width": "0%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#E90000",
          "width": "0%",
        }
      }
    />
  </div>
  <svg
    className="c5"
    height="225.25"
    width="16.666666666666668"
  >
    <polygon
      className="c6"
      points="0,74 0,225.25 16.666666666666668,225.25 16.666666666666668,0"
    />
    <polygon
      className="c7"
      points="0,0 0,74 16.666666666666668,0"
    />
  </svg>
</div>
`;

exports[`<SteadyBar> renders with Tooltip when selected 1`] = `
.c0 {
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c2 {
  border-top: 1px dotted gray;
  z-index: 5;
}

.c3 {
  position: absolute;
  bottom: 0px;
  left: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4:first-child {
  border-bottom-left-radius: 5px;
}

.c4:last-child {
  border-bottom-right-radius: 5px;
}

.c5 {
  position: absolute;
  bottom: 0;
  left: 0;
}

.c6 {
  fill: transparent;
}

.c7 {
  fill: white;
  display: hidden;
}

<div
  className="c0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "zIndex": 1,
    }
  }
>
  <Tooltip
    interval={
      Object {
        "cadence": undefined,
        "endIntensity": 0.901,
        "id": "mock-id",
        "length": Object {
          "seconds": 50,
        },
        "pace": 0,
        "startIntensity": 0.605,
        "type": "ramp",
      }
    }
    mode={
      BikeMode {
        "ftp": 250,
        "lengthType": "time",
        "sportType": "bike",
        "weight": 75,
      }
    }
    onCadenceChange={[Function]}
  />
  <div
    className="c1"
    onClick={[Function]}
  >
    <Resizable
      className="c2"
      enable={
        Object {
          "right": true,
          "top": true,
        }
      }
      grid={
        Array [
          1,
          1,
        ]
      }
      maxHeight={500}
      minHeight={25}
      minWidth={3}
      onResize={[Function]}
      onResizeStop={[Function]}
      size={
        Object {
          "height": 151.25,
          "width": 8.333333333333334,
        }
      }
    />
    <Resizable
      className="c2"
      enable={
        Object {
          "right": true,
          "top": true,
        }
      }
      grid={
        Array [
          1,
          1,
        ]
      }
      maxHeight={500}
      minHeight={25}
      minWidth={3}
      onResize={[Function]}
      onResizeStop={[Function]}
      size={
        Object {
          "height": 225.25,
          "width": 8.333333333333334,
        }
      }
    />
  </div>
  <div
    className="c3"
    style={
      Object {
        "backgroundColor": "#0E90D4",
        "flexDirection": "row",
        "height": 225.25,
      }
    }
  >
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#807F80",
          "width": "0%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#0E90D4",
          "width": "53.37837837837838%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#00C46A",
          "width": "46.62162162162162%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#FFCB00",
          "width": "0%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#FF6430",
          "width": "0%",
        }
      }
    />
    <div
      className="c4"
      style={
        Object {
          "backgroundColor": "#E90000",
          "width": "0%",
        }
      }
    />
  </div>
  <svg
    className="c5"
    height="225.25"
    width="16.666666666666668"
  >
    <polygon
      className="c6"
      points="0,74 0,225.25 16.666666666666668,225.25 16.666666666666668,0"
    />
    <polygon
      className="c7"
      points="0,0 0,74 16.666666666666668,0"
    />
  </svg>
</div>
`;
